connectivity/FEATURE_BLE/source/cordio/source/GattServerImpl.cpp

```diff
@@ -1146,8 +1148,23 @@ uint8_t GattServer::atts_read_cb(
             return read_auth_params.authorizationReply & 0xFF;
         }
 
-        pAttr->pValue = read_auth_params.data;
-        *pAttr->pLen = read_auth_params.len;
+        /* if new data provided copy into the attribute value buffer */
+        if (read_auth_params.data) {
+            if (read_auth_params.len > pAttr->maxLen) {
+                tr_error("Read authorisation callback set length larger than maximum attribute length, "
+                         "cannot copy data");
+                err = ATT_ERR_UNLIKELY;
+            } else {
+                memcpy(pAttr->pValue, read_auth_params.data, read_auth_params.len);
+                *pAttr->pLen = read_auth_params.len;
+
+                if (read_auth_params.len < offset) {
+                    tr_warning("Read authorisation callback shortened data beyond current offset, "
+                               "current read will fail");
+                    err = ATT_ERR_OFFSET;
```

Comment:
I think this should be done before copying anything in the attribute value.

Reply:
I think it's OK to update the value - it's the least unexpected thing. What if the user relies on the copy to happen because that data will be gone later.

Reply:
I think you should use `GattServer::write` for such cases. That's fine, we can keep it that way.